tracing/http.go

```diff
@@ -0,0 +1,239 @@
+
+	//RequestPayloadAnnotationKey is a key under which request payload stored in span
+	RequestPayloadAnnotationKey = "request.payload"
+
+	//ResponsePayloadAnnotationKey is a key under which response payload stored in span
+	ResponsePayloadAnnotationKey = "response.payload"
+
+	//DefaultMaxPayloadSize represent max payload size which will be added to span
+	DefaultMaxPayloadSize = 1024 * 1024
+
+	//TruncatedMarkerKey is a key for annotation which will be presented in span in case payload was truncated
+	TruncatedMarkerKey = "payload.truncated"
+
+	//TruncatedMarkerValue is a value for annotation which will be presented in span in case payload was truncated
+	TruncatedMarkerValue = "true"
+)
+
+type headerMatcher func(string) (string, bool)
+
+type httpOptions struct {
+	spanWithHeaders func(*http.Request) bool
+	headerMatcher   headerMatcher
+
+	spanWithPayload func(*http.Request) bool
+	maxPayloadSize  int
+}
+
+//HTTPOption allows extending handler with additional functionality
+type HTTPOption func(*httpOptions)
+
+func defaultHTTPOptions() *httpOptions {
+	return &httpOptions{
+		headerMatcher:  defaultHeaderMatcher,
+		maxPayloadSize: DefaultMaxPayloadSize,
+
+		//Keep spanWithHeaders and spanWithPayload equals to nil instead of dummy functions
+		//to prevent path trough header for each request
+	}
+}
+
+//WithHeadersAnnotation annotate span with http headers
+func WithHeadersAnnotation(f func(*http.Request) bool) HTTPOption {
+	return func(ops *httpOptions) {
+		ops.spanWithHeaders = f
+	}
+}
+
+//WithHeaderMatcher set header matcher to filterout or preprocess headers
+func WithHeaderMatcher(f func(string) (string, bool)) HTTPOption {
+	return func(ops *httpOptions) {
+		ops.headerMatcher = f
+	}
+}
+
+//WithPayloadAnnotation add request/response body as an attribute to span if f returns true
+func WithPayloadAnnotation(f func(*http.Request) bool) HTTPOption {
+	return func(ops *httpOptions) {
+		ops.spanWithPayload = f
+	}
+}
+
+//WithHTTPPayloadSize limit payload size propogated to span
+//in case payload exceeds limit, payload truncated and
+//annotation payload.truncated=true added into span
+func WithHTTPPayloadSize(maxSize int) HTTPOption {
+	return func(ops *httpOptions) {
+		ops.maxPayloadSize = maxSize
+	}
+}
+
+//NewMiddleware wrap handler
+func NewMiddleware(ops ...HTTPOption) func(http.Handler) http.Handler {
+	options := defaultHTTPOptions()
+	for _, op := range ops {
+		op(options)
+	}
+
+	return func(h http.Handler) http.Handler {
+		och := &ochttp.Handler{
+			Handler: &Handler{
+				child:   h,
+				options: options,
+			},
+		}
+
+		return och
+	}
+}
+
+//Check that &Handler comply with http.Handler interface
+var _ http.Handler = &Handler{}
+
+//Handler is a opencensus http plugin wrapper which do some usefull things to reach traces
+type Handler struct {
+	child http.Handler
+
+	options *httpOptions
+}
+
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
```

Comment:
What use case is this covering? If I make a grpc call to the service, it won't run this code. Wouldn't it be simpler to handle the gRPC case and not bother implementing the http Handler. We only need gateway to pass along headers to the underlying gRPC server

Reply:
You are correct, however we want to be able to trace the call flow at all the protocol boundaries for a service.  This allows us to not only see what the call properties are (headers, payload, etc) but we can also collect timing associated with the call pipeline for any entry point into/out of the service.